Extension/src/LanguageServer/Providers/CopilotHoverProvider.ts

```diff
@@ -0,0 +1,149 @@
+import * as nls from 'vscode-nls';
+import { DefaultClient, GetCopilotHoverInfoParams, GetCopilotHoverInfoRequest } from '../client';
+import { RequestCancelled, ServerCancelled } from '../protocolFilter';
+import { CppSettings } from '../settings';
+
+nls.config({ messageFormat: nls.MessageFormat.bundle, bundleFormat: nls.BundleFormat.standalone })();
+const localize: nls.LocalizeFunc = nls.loadMessageBundle();
+
+export class CopilotHoverProvider implements vscode.HoverProvider {
+    private client: DefaultClient;
+    private currentDocument: vscode.TextDocument | undefined;
+    private currentPosition: vscode.Position | undefined;
+    private currentCancellationToken: vscode.CancellationToken | undefined;
+    private waiting: boolean = false;
+    private ready: boolean = false;
+    private cancelled: boolean = false;
+    private cancelledDocument: vscode.TextDocument | undefined;
+    private cancelledPosition: vscode.Position | undefined;
+    private content: string | undefined;
+    constructor(client: DefaultClient) {
+        this.client = client;
+    }
+
+    public async provideHover(document: vscode.TextDocument, position: vscode.Position, token: vscode.CancellationToken): Promise<vscode.Hover | undefined> {
+        await this.client.ready;
+
+        const settings: CppSettings = new CppSettings(vscode.workspace.getWorkspaceFolder(document.uri)?.uri);
+        if (settings.hover === "disabled") {
+            return undefined;
+        }
+
+        const newHover = this.isNewHover(document, position);
+        if (newHover) {
+            this.reset();
+        }
+
+        // Wait for the main hover provider to finish and confirm it has content.
+        const hoverProvider = this.client.getHoverProvider();
+        if (!await hoverProvider?.contentReady) {
+            return undefined;
+        }
+
+        if (token.isCancellationRequested) {
+            throw new vscode.CancellationError();
+        }
+        this.currentCancellationToken = token;
+
+        if (!newHover) {
+            if (this.ready) {
+                const contentMarkdown = new vscode.MarkdownString(`$(sparkle) Copilot\n\n${this.content}`, true);
+                return new vscode.Hover(contentMarkdown);
+            }
+            if (this.waiting) {
+                const loadingMarkdown = new vscode.MarkdownString("$(sparkle) $(loading~spin)", true);
+                return new vscode.Hover(loadingMarkdown);
+            }
+        }
+
+        this.currentDocument = document;
+        this.currentPosition = position;
+        const commandString = "$(sparkle) [" + localize("generate.copilot.description", "Generate Copilot summary") + "](command:C_Cpp.ShowCopilotHover \"" + localize("copilot.disclaimer", "AI-generated content may be incorrect.") + "\")";
+        const commandMarkdown = new vscode.MarkdownString(commandString);
+        commandMarkdown.supportThemeIcons = true;
+        commandMarkdown.isTrusted = true;
```

Comment:
It's not a big risk in this case since there's no user-generated content in the first hover, but for security prefer setting `isTrusted` to an exact set of enabled commands instead of boolean.

```suggestion
        commandMarkdown.isTrusted = { enabledCommands: ["C_Cpp.ShowCopilotHover"] };
```